### DISCUSSION TITLE
when i use dittoclient, it will disconnect at every  10 minutes ,why?

### DISCUSSION THREAD

**Discussion by timyang9527** [eclipse-ditto/ditto#1566] [Q&A]
when i use dittoclient, it will disconnect at every  10 minutes ,why?

**use client** 
 client.live().forId(thingId)
                        .message()
                        .to()
                        .subject("routercmd")
                        .timeout(Duration.ofSeconds(60))
                        .payload(JSONObject.toJSONString(requestBody))
                        .send(String.class, (response, throwable) -> {
                            future.complete(response.getPayload().orElse(null));
                        });
                String s = future.get(3000, TimeUnit.MILLISECONDS);
                logger.info("call result ={}" ,s);
                RouterResponseBody responseBody = JSONObject.parseObject(s, RouterResponseBody.class);

 **create client** 
  public DittoClient create() throws Exception {
        log.info("start create ditto client");
        return DittoClients.newInstance(createMessagingProvider())
                .connect()
                .toCompletableFuture()
                .join();

    public static MessagingProvider createMessagingProvider() {
        final MessagingConfiguration.Builder builder = WebSocketMessagingConfiguration.newBuilder()
                .endpoint(DITTO_ENDPOINT_URL)
                .jsonSchemaVersion(JsonSchemaVersion.V_2)
                .reconnectEnabled(false);

        final ProxyConfiguration proxyConfiguration;
        if (PROXY_HOST != null && !PROXY_HOST.isEmpty()) {
            proxyConfiguration = ProxyConfiguration.newBuilder()
                    .proxyHost(PROXY_HOST)
                    .proxyPort(Integer.parseInt(PROXY_PORT))
                    .build();
            builder.proxyConfiguration(proxyConfiguration);
        } else {
            proxyConfiguration = null;
        }

        if (DITTO_TRUSTSTORE_LOCATION != null) {
            builder.trustStoreConfiguration(TrustStoreConfiguration.newBuilder()
                    .location(DITTO_TRUSTSTORE_LOCATION)
                    .password(DITTO_TRUSTSTORE_PASSWORD)
                    .build());
        }

        final AuthenticationProvider<WebSocket> authenticationProvider;
        if (DITTO_DUMMY_AUTH_USER != null) {
            authenticationProvider =
                    AuthenticationProviders.dummy(DummyAuthenticationConfiguration.newBuilder()
                            .dummyUsername(DITTO_DUMMY_AUTH_USER)
                            .build());
        } else if (DITTO_OAUTH_CLIENT_ID != null && !DITTO_OAUTH_CLIENT_ID.isEmpty()) {
            final ClientCredentialsAuthenticationConfiguration.ClientCredentialsAuthenticationConfigurationBuilder
                    authenticationConfigurationBuilder =
                    ClientCredentialsAuthenticationConfiguration.newBuilder()
                            .clientId(DITTO_OAUTH_CLIENT_ID)
                            .clientSecret(DITTO_OAUTH_CLIENT_SECRET)
                            .scopes(DITTO_OAUTH_SCOPES)
                            .tokenEndpoint(DITTO_OAUTH_TOKEN_ENDPOINT);
            if (proxyConfiguration != null) {
                authenticationConfigurationBuilder.proxyConfiguration(proxyConfiguration);
            }
            authenticationProvider =
                    AuthenticationProviders.clientCredentials(authenticationConfigurationBuilder.build());
        } else {
            authenticationProvider =
                    AuthenticationProviders.basic(BasicAuthenticationConfiguration.newBuilder()
                            .username(DITTO_USERNAME)
                            .password(DITTO_PASSWORD)
                            .build());
        }

        return MessagingProviders.webSocket(builder.build(), authenticationProvider);
    }

**Comment by thjaeckle** [comment]
Hi @AriesMg 
This sounds like e.g. your infrastructure (e.g. a HTTP reverse proxy) could have a max idle timeout of 10 minutes for open connections.
On which infrastructure do you operate Ditto and which loadbalancers / reverse proxies are in front of Ditto?

**Reply by timyang9527** [reply]
use  nginx  as  loadbalancers

**Reply by thjaeckle** [reply]
And did you configure the `"/ws/"` resource of Ditto accordingly in your nginx config?
There are many resources which explain how to do that, e.g. https://mpolinowski.github.io/docs/Development/Javascript/2021-09-09--websocket-NGINX/2021-09-09/

**Reply by timyang9527** [reply]
i want control Thousands Of router via dittclient，how many dittclients should I create on my server？ just 1 enough or need more?

**Comment by timyang9527** [comment]
@thjaeckle i want control Thousands Of router via dittclient，how many dittclients should I create on my server？ just 1 enough or need more?

**Reply by thjaeckle** [reply]
That depends on your use case - which I know nothing about - and mainly amount of messages sent/received.

**Reply by timyang9527** [reply]
**thanks,
 i have a new qustion ，every 10 minutes, i will find a mount of error API  error ,What is the most likely cause**
2023-01-19 10:20:08 [ReadingThread] DEBUG org.eclipse.ditto.client.messaging.internal.WebSocketMessagingProvider:553] Client <nginx:ditto:5920353c-b59f-4d02-a63c-37ce5d547bce>: Received WebSocket string message
 <{"topic":"routerX/MWEWYZM5NJQWMZKW/things/live/messages/routercmd","headers":{"response-required":false,"ditto-connection-id":"mqtt-connection-id","ditto-originator":"nginx:ditto","version":2,"channel":"live",
"timeout":"60000ms","correlation-id":"537d6f0a-5655-4ddc-af59-7daa56465bc2","content-type":"text/plain","ditto-message-thing-id":"routerX:MWEWYZM5NJQWMZKW","ditto-message-subject":"routercmd","ditto-message-dire
ction":"FROM","status":"500"},"path":"/outbox/messages/routercmd","value":"API error","status":500}>

**Reply by thjaeckle** [reply]
Your device seems to send this error with `"value": "API error"` (Ditto does not create such meaningless error messages) - Ditto just routes that response error to the original sender of the message.

**Comment by thjaeckle** [comment]
As there are only new (unrelated) questions and not a feedback on my questions and suggestions, I will close this discsussion (as this is not a discussion).
